# A Quick Tour of Variables and Data Types in Python

This tutorial is a part of [Data Analysis with Python: Zero to Pandas](https://jovian.ai/learn/data-analysis-with-python-zero-to-pandas) and [Zero to Data Analyst Science Bootcamp](https://jovian.ai/learn/zero-to-data-analyst-bootcamp).


![](https://i.imgur.com/6cg2E9Q.png)
These tutorials take a practical and coding-focused approach. The best way to learn the material is to execute the code and experiment with it yourself.

This tutorial covers the following topics:

- Storing information using variables
- Primitive data types in Python: Integer, Float, Boolean, None and String
- Built-in data structures in Python: List, Tuple and Dictionary
- Methods and operators supported by built-in data types

## Storing information using variables

Computers are useful for two purposes: storing information (also known as data) and performing operations on stored data. While working with a programming language such as Python, data is stored in variables. You can think of variables are containers for storing data. The data stored within a variable is called its value. Creating variables in Python is pretty easy, as we've already seen in the [previous tutorial](https://jovian.ml/aakashns/first-steps-with-python/v/4#C15).



In [13]:
# Favorite colors

my_favorite_color = "blue"

In [14]:
my_favorite_color

'blue'

A variable is created using an assignment statement. It begins with the variable's name, followed by the assignment operator `=` followed by the value to be stored within the variable.  Note that the assignment operator `=` is different from the equality comparison operator `==`.

You can also assign values to multiple variables in a single statement by separating the variable names and values with commas.

In [15]:
color1, color2, color3 = "red", "green", "blue"

In [16]:
color1

'red'

In [17]:
color2

'green'

In [18]:
color3

'blue'

You can assign the same value to multiple variables by chaining multiple assignment operations within a single statement.

In [19]:
color4 = color5 = color6 = "magenta"

In [20]:
color4

'magenta'

In [21]:
color5

'magenta'

In [22]:
color6

'magenta'

You can change the value stored within a variable by assigning a new value to it using another assignment statement. Be careful while reassigning variables: when you assign a new value to the variable, the old value is lost and no longer accessible.

In [23]:
my_favorite_color = "red"

In [24]:
my_favorite_color

'red'

While reassigning a variable, you can also use the variable's previous value to compute the new value.

In [25]:
counter = 10

In [26]:
counter

10

In [27]:
counter = 11

In [28]:
counter = counter + 1

In [29]:
counter

12

The pattern `var = var op something` (where `op` is an arithmetic operator like `+`, `-`, `*`, `/`) is very common, so Python provides a *shorthand* syntax for it.

In [30]:
counter = 10

In [31]:
counter = counter + 4

In [32]:
counter

14

In [33]:
# Same as `counter = counter + 4`
counter += 4

In [34]:
counter

18

In [35]:
counter += 4

In [136]:
counter

22

### Documentation for the `counter` variable

This cell displays the current value of the `counter` variable. The `counter` variable is used throughout the notebook to demonstrate arithmetic operations and assignment shorthands. Its value has been incremented multiple times in the preceding cells.

Variable names can be short (`a`, `x`, `y`, etc.) or descriptive ( `my_favorite_color`, `profit_margin`, `the_3_musketeers`, etc.). However, you must follow these rules while naming Python variables:

* A variable's name must start with a letter or the underscore character `_`. It cannot begin with a number.
* A variable name can only contain lowercase (small) or uppercase (capital) letters, digits, or underscores (`a`-`z`, `A`-`Z`, `0`-`9`, and `_`).
* Variable names are case-sensitive, i.e., `a_variable`, `A_Variable`, and `A_VARIABLE` are all different variables.

Here are some valid variable names:

In [ ]:
a_variable = 23
is_today_Saturday = False
my_favorite_car = "Delorean"
the_3_musketeers = ["Athos", "Porthos", "Aramis"]

Let's try creating some variables with invalid names. Python prints a syntax error if your variable's name is invalid.

> **Syntax**: The syntax of a programming language refers to the rules that govern the structure of a valid instruction or *statement*. If a statement does not follow these rules, Python stops execution and informs you that there is a *syntax error*. You can think of syntax as the rules of grammar for a programming language.

In [141]:
# SyntaxError: Variable names cannot start with a number.
2_day = 555

SyntaxError: invalid decimal literal (995123115.py, line 2)

This cell throws a `SyntaxError` because variable names in Python cannot start with a number. Valid variable names must begin with a letter or an underscore (`_`).

In [185]:
# SyntaxError: Variable names can only contain letters, digits, or underscores. The '@' symbol is not allowed.
so@oi = '7848'

SyntaxError: cannot assign to expression here. Maybe you meant '==' instead of '='? (408918180.py, line 2)

This cell also results in a `SyntaxError` because variable names can only contain letters, digits, or underscores. The `@` symbol is not allowed in variable names.

## Built-in data types in Python

Any data or information stored within a Python variable has a *type*. You can view the type of data stored within a variable using the `type` function.

In [ ]:
my_age = 29

In [ ]:
my_favorite_car = "Lamborgini"

In [ ]:
is_today_Saturday = True

In [ ]:
the_3_learners = ["Reuben", "Victor", "Favour"]

In [ ]:
my_age

In [ ]:
type(my_age)

The `type()` function in Python is used to determine the data type of an object. Here, it confirms that `my_age` (which holds `29`) is an integer (`int`).

In [ ]:
type(is_today_Saturday)

This cell uses `type()` to show that `is_today_Saturday` (which holds `True`) is a boolean (`bool`).

In [ ]:
my_favorite_car

In [ ]:
type(my_favorite_car)

Here, `type()` reveals that `my_favorite_car` (a string like 'Lamborghini') is of type `str`.

In [ ]:
the_3_learners

In [ ]:
type(the_3_learners)

The `type()` function confirms that `the_3_learners` is a list (`list`).

In [ ]:
the_seed_learners = ('Marvelous', 'Akeem', 'Jennifer', "Hero", "Lovelyn")

In [ ]:
the_seed_learners

In [ ]:
type(the_seed_learners)

This cell verifies that `the_seed_learners` is a tuple (`tuple`), which is an ordered, immutable collection of items.

In [ ]:
print(the_seed_learners)

Python has several built-in data types for storing different kinds of information in variables. Following are some commonly used data types:

1. Integer
2. Float
3. Boolean
4. None
5. String
6. List
7. Tuple
8. Dictionary

Integer, float, boolean, None, and string are *primitive data types* because they represent a single value. Other data types like list, tuple, and dictionary are often called *data structures* or *containers* because they hold multiple pieces of data together.

### Integer

Integers represent positive or negative whole numbers, from negative infinity to infinity.
`Note that integers should not include decimal points`.

Integers have the type `int`.

In [173]:
current_year = 2024

In [174]:
# This variable exists and is an integer.
current_year

2024

In [ ]:
type(current_year)

Confirming `current_year` is an integer (`int`).

Unlike some other programming languages, integers in Python can be arbitrarily large (or small). There's no lowest or highest value for integers, and there's just one `int` type (as opposed to `short`, `int`, `long`, `long long`, `unsigned int`, etc. in C/C++/Java).

In [214]:
a_large_negative_number = -23374038374832934334234317348343

In [215]:
a_large_negative_number

-23374038374832934334234317348343

In [216]:
type(a_large_negative_number)

int

This shows that Python integers can handle arbitrarily large numbers, maintaining their `int` type.

### Float

Floats (or floating-point numbers) are numbers with a decimal point. There are no limits on the value or the number of digits before or after the decimal point. Floating-point numbers have the type `float`.

In [211]:
pi = 3.141592653589793238

In [212]:
pi

3.141592653589793

In [213]:
type(pi)

float

The `type()` function confirms that `pi` is a float (`float`).

Note that a whole number is treated as a float if written with a decimal point, even though the decimal portion of the number is zero.

In [180]:
a_number = 3.

In [181]:
a_number

3.0

In [186]:
# NameError: 'a_number' was defined as 3. earlier, but if run in isolation without that cell, it fails.
type(a_number)

float

This cell produced a `NameError` because `a_number` was not defined or executed in the current session context when this cell was run. If `a_number = 3.` was executed, it would show `float`.

This cell produced a `NameError` because `a_number` was not defined or executed in the current session context when this cell was run. If `a_number = 3.` was executed, it would show `float`.

In [194]:
another_number = 4.0

In [196]:
another_number

4.0

In [200]:
type(another_number)

float

Confirming `another_number` as a float. Even with `.0`, Python recognizes it as a float.

Floating point numbers can also be written using the scientific notation with an "e" to indicate the power of 10.

In [218]:
one_hundredth = 1e-2

In [219]:
one_hundredth

0.01

In [220]:
type(one_hundredth)

float

Demonstrates `one_hundredth` is a `float`, even when defined using scientific notation.

In [217]:
one_hundred = 1e2

In [ ]:
one_hundred

In [ ]:
type(one_hundred)

Demonstrates `one_hundred` is a `float`, even when defined using scientific notation.

In [221]:
type(one_hundredth)

float

In [222]:
avogadro_number = 6.02214076e23

In [223]:
avogadro_number

6.02214076e+23

In [224]:
type(avogadro_number)

float

Confirms `avogadro_number` as a float due to its scientific notation.

In [ ]:
num = 1e2

In [ ]:
num

In [ ]:
type(num)

Confirms `num` as a float when using scientific notation.

**You can convert floats into integers and vice versa using the `float` and `int` functions.**

**The operation of converting one type of value into another is called `casting`.**

In [205]:
current_year

2024

In [206]:
float(current_year)

2024.0

In [207]:
current_year = float(current_year)

In [203]:
current_year

2024

In [204]:
# NameError: This fails if the cell defining 'current_year' hasn't been executed in the current session.
current_year

2024

This cell also resulted in a `NameError` because `current_year` was not defined in the current session. The notebook kernel needs to have `current_year` assigned a value for this cell to execute successfully.

In [201]:
type(current_year)

int

In [202]:
# NameError: Attempting to check the type of 'current_year' which might not be initialized.
type(current_year)

int

Similar to the previous errors, this cell failed with a `NameError` because `current_year` was not defined in the current execution environment. To check its type, `current_year` must first be assigned a value.

After casting `current_year` to a float, its type is now `float`.

In [225]:
a_large_negative_number

-23374038374832934334234317348343

This cell displays the `a_large_negative_number` variable, which was previously defined as a large integer.

In [226]:
float(a_large_negative_number)

-2.3374038374832935e+31

Converting the large integer `a_large_negative_number` to a `float` results in a floating-point representation, often using scientific notation for very large or small numbers.

In [227]:
# NameError: 'pi' was not defined prior to this cell being run.
pi

3.141592653589793

This cell produced a `NameError` because `pi` was not defined in the current session context when this cell was run. If `pi` were defined, this would attempt to convert it to an integer, truncating the decimal part.

This cell produced a `NameError` because `pi` was not defined in the current session context when this cell was run. If `pi` were defined, this would attempt to convert it to an integer, truncating the decimal part.

In [228]:
# NameError: Cannot cast 'pi' to an integer because 'pi' was not defined in this session.
int(pi)

3

This cell resulted in a `NameError` because the variable `pi` was not defined in the current execution environment before attempting to convert it to an integer. You need to ensure `pi` is assigned a value first (e.g., `pi = 3.14159...`).

In [229]:
avogadro_number

6.02214076e+23

This cell displays the `avogadro_number` variable, which is a float represented in scientific notation.

In [230]:
int(avogadro_number)

602214075999999987023872

Converting the float `avogadro_number` to an integer using `int()` truncates the decimal part, effectively rounding down towards zero.

In [231]:
int(3456.0)

3456

This demonstrates that converting a float literal (even one with a `.0` decimal) to an integer simply removes the decimal part.

While performing arithmetic operations, integers are automatically converted to `float`s if any of the operands is a `float`. Also, the division operator `/` always returns a `float`, even if both operands are integers. Use the `//` operator if you want the result of the division to be an `int`.

In [232]:
45 + 3.0

48.0

When performing an arithmetic operation (like addition) between an integer and a float, Python automatically converts the integer to a float, and the result is a float.

In [233]:
type(45 * 3.0)

float

This confirms that the result of multiplying an integer and a float is a `float`.

In [234]:
45 * 3.

135.0

Multiplying an integer by a float (even if the float is represented with a `.0`) results in a float.

In [235]:
type(45 * 3)

int

This confirms that multiplying two integers results in an `int`.

In [236]:
10/2

5.0

In Python 3, the division operator `/` always returns a `float`, even if the result is a whole number.

In [237]:
type(10/3)

float

This explicitly shows that the result type of standard division (`/`) is `float`.

In [238]:
10/4

2.5

Another example of standard division, resulting in a float.

In [239]:
10//2

5

The integer division operator `//` performs division and truncates the result to an integer (rounds down to the nearest whole number).

In [39]:
type(10/2)

float

This confirms the type of the result from standard division (`/`) is `float`.

In [40]:
10 % 3

1

The modulo operator `%` returns the remainder of a division.

In [41]:
type(10//2)

int

This confirms the type of the result from integer division (`//`) is `int`.

In [42]:
round(23.67798)

24

The `round()` function rounds a floating-point number to the nearest integer. By default, it rounds to the nearest even number if the fractional part is exactly 0.5.

In [43]:
round(23.67798, 2)

23.68

The `round()` function can also take a second argument to specify the number of decimal places to round to.

In [44]:
import math

This cell imports the `math` module, which provides access to a set of mathematical functions.

In [45]:
math.ceil(23.2798)

24

The `math.ceil()` function returns the smallest integer greater than or equal to the given number (rounds up).

In [46]:
math.floor(23.2798)

23

The `math.floor()` function returns the largest integer less than or equal to the given number (rounds down).

In [47]:
round(23.27798)

23

Another example of `round()`, demonstrating that it rounds to the nearest integer. Here, since .27798 is less than .5, it rounds down.

### Boolean

Booleans represent one of 2 values: `True` and `False`. Booleans have the type `bool`.

In [48]:
type(False)

bool

This confirms that the literal `False` is of type `bool`.

In [49]:
type(True)

bool

This confirms that the literal `True` is of type `bool`.

In [50]:
is_today_Sunday = False

Here, `is_today_Sunday` is assigned the boolean value `False`.

In [51]:
is_today_Sunday

False

Printing the value of `is_today_Sunday` confirms it holds `False`.

In [52]:
is_today_Saturday = True

Here, `is_today_Saturday` is assigned the boolean value `True`.

In [53]:
is_today_Saturday

True

Printing the value of `is_today_Saturday` confirms it holds `True`.

In [54]:
type(is_today_Saturday)

bool

Confirming that the variable `is_today_Saturday` is of type `bool`.

Booleans are generally the result of a comparison operation, e.g., `==`, `>=`, etc.

In [55]:
cost_of_ice_bag = 1.25


is_ice_bag_expensive = cost_of_ice_bag >= 10

This cell assigns a float value to `cost_of_ice_bag` and then uses a comparison operator (`>=`) to assign a boolean value to `is_ice_bag_expensive` based on whether the cost is greater than or equal to 10.

In [56]:
is_ice_bag_expensive

False

The `is_ice_bag_expensive` variable holds `False` because `1.25` is not greater than or equal to `10`.

In [57]:
type(is_ice_bag_expensive)

bool

This confirms that the result of a comparison operation, stored in `is_ice_bag_expensive`, is of type `bool`.

Booleans are automatically converted to `int`s when used in arithmetic operations. `True` is converted to `1` and `False` is converted to `0`.

In [58]:
5 + False ## False = 0

5

When `False` (which is `0`) is used in an arithmetic operation with an integer, it is implicitly converted to `0`. So, `5 + 0` results in `5`.

In [59]:
3.0 + True  ### True = 1

4.0

Similarly, when `True` (which is `1`) is used in an arithmetic operation with a float, it is implicitly converted to `1`. The result `3.0 + 1` is `4.0` (a float).

In [60]:
type(10000)

int

This demonstrates checking the type of an integer literal directly.

In [61]:
hgu = bool()
hgu

False

This initializes `hgu` with the default boolean value, `False`, which is returned when `bool()` is called with no arguments. It then displays `hgu`.

In [62]:
type(hgu)

bool

This confirms that an empty `bool()` constructor returns `False`, and its type is `bool`.

In [63]:
"Solomon"

'Solomon'

This cell evaluates the string literal 'Solomon' and prints it as output.

In [64]:
bool(())

False

This shows that an empty tuple `()` evaluates to `False` when converted to a boolean.

Any value in Python can be converted to a Boolean using the `bool` function.

Only the following values evaluate to `False` (they are often called *falsy* values):

1. The value `False` itself
2. The integer `0`
3. The float `0.0`
4. The empty value `None`
5. The empty text `""`
6. The empty list `[]`
7. The empty tuple `()`
8. The empty dictionary `{}`
9. The empty set `set()`
10. The empty range `range(0)`

Everything else evaluates to `True` (a value that evaluates to `True` is often called a *truthy* value).

In [65]:
bool(False)

False

This explicitly converts `False` to a boolean, which remains `False`.

In [66]:
bool(0)

False

The integer `0` is a falsy value, so `bool(0)` returns `False`.

In [67]:
bool(0.0)

False

The float `0.0` is also a falsy value, so `bool(0.0)` returns `False`.

In [68]:
bool(None)

False

The `None` object is a falsy value, so `bool(None)` returns `False`.

In [69]:
bool("")

False

An empty string `""` is a falsy value, so `bool("")` returns `False`.

In [70]:
bool([])

False

An empty list `[]` is a falsy value, so `bool([])` returns `False`.

In [71]:
bool(())

False

An empty tuple `()` is a falsy value, so `bool(())` returns `False`.

In [72]:
bool({})

False

An empty dictionary `{}` is a falsy value, so `bool({})` returns `False`.

In [73]:
bool(set())

False

An empty set `set()` is a falsy value, so `bool(set())` returns `False`.

In [74]:
bool(range(0))

False

An empty range `range(0)` is a falsy value, so `bool(range(0))` returns `False`.

In [75]:
bool(True), bool(1), bool(2000.0), bool("hello"), bool([1,2]), bool((2,3)), bool(range(10))

(True, True, True, True, True, True, True)

This demonstrates that various non-empty values (True, non-zero numbers, non-empty strings, non-empty lists, non-empty tuples, non-empty ranges) are all truthy, meaning `bool()` converts them to `True`.

In [76]:
bool(0)

False

Just like the earlier `bool(0)`, this confirms that integer `0` is a falsy value.

### None

The None type includes a single value `None`, used to indicate the absence of a value. `None` has the type `NoneType`. It is often used to declare a variable whose value may be assigned later.

In [77]:
Nothing = None

The variable `Nothing` is assigned the `None` object, indicating an absence of value.

In [78]:
nothing = None

The variable `nothing` is also assigned the `None` object.

In [79]:
print(nothing)

None


Printing `nothing` explicitly outputs the word 'None'.

In [80]:
type(nothing)

NoneType

This confirms that the `None` object has its own unique type: `NoneType`.

### String

A string is used to represent text (*a string of characters*) in Python. Strings must be surrounded using quotations (either the single quote `'` or the double quote `"`). Strings have the type `string`.

In [81]:
'Monday'

'Monday'

In [82]:
"Monday"

'Monday'

In [241]:
# SyntaxError: The string starts with a single quote but ends with a double quote, causing an unterminated string literal.
"Monday"

'Monday'

This cell throws a `SyntaxError` because the string starts with a single quote but ends with a double quote, leading to an unterminated string literal.

In [242]:
today = "Sunday"

In [243]:
today

'Sunday'

In [244]:
type(today)

str

This confirms `today` as a string (`str`).

In [245]:
# NameError: The variable 'current_year' was not defined before attempting to convert its type.
float(current_year)

2024.0

This cell produced a `NameError` because `current_year` was not defined or executed in the current session context when this cell was run. You need to ensure the variable is defined before attempting to convert its type.

You can use single quotes inside a string written with double quotes, and vice versa.

In [246]:
my_favorite_movie = "One Flew over the Cuckoo's Nest"

In [247]:
my_favorite_movie

"One Flew over the Cuckoo's Nest"

In [248]:
my_favorite_pun = 'Thanks for explaining the word "many" to me, it means a lot.'

In [249]:
my_favorite_pun

'Thanks for explaining the word "many" to me, it means a lot.'

To use a double quote within a string written with double quotes, *escape* the inner quotes by prefixing them with the `\` character.

In [250]:
another_pun = "The first time I got a universal remote control, I thought to myself \"This changes everything\"."

In [251]:
another_pun

'The first time I got a universal remote control, I thought to myself "This changes everything".'

Strings created using single or double quotes must begin and end on the same line. To create multiline strings, use three single quotes `'''` or three double quotes `"""` to begin and end the string. Line breaks are represented using the newline character `\n`.

In [252]:
yet_another_pun = '''Son: "Dad, can you tell me what a solar eclipse is?"
Dad: "No sun."
"Sunny" '''

In [ ]:
yet_another_pun

In [253]:
print(yet_another_pun)

Son: "Dad, can you tell me what a solar eclipse is?"
Dad: "No sun."
"Sunny" 


In [254]:
print('Dad: \n"No sun."')

Dad: 
"No sun."


Multiline strings are best displayed using the `print` function.

In [255]:
print(yet_another_pun)

Son: "Dad, can you tell me what a solar eclipse is?"
Dad: "No sun."
"Sunny" 


In [256]:
a_music_pun = '''
Two windmills are standing in a field and one asks the other,
"What kind of music do you like?"

The other says,
"I'm a big metal fan."
'''

In [258]:
a_music_pun

'\nTwo windmills are standing in a field and one asks the other,\n"What kind of music do you like?"\n\nThe other says,\n"I\'m a big metal fan."\n'

In [257]:
print(a_music_pun)


Two windmills are standing in a field and one asks the other,
"What kind of music do you like?"

The other says,
"I'm a big metal fan."



In [259]:
help(len)

Help on built-in function len in module builtins:

len(obj, /)
    Return the number of items in a container.



In [260]:
len(a_music_pun)

137

You can check the length of a string using the `len` function.

In [261]:
my_favorite_movie

"One Flew over the Cuckoo's Nest"

In [262]:
len(my_favorite_movie)

31

Note that special characters like `\n` and escaped characters like `\"` count as a single character, even though they are written and sometimes printed as two characters.

In [263]:
empty = '    '
len(empty)

4

In [264]:
multiline_string = """a
b"""
print(multiline_string)

a
b


In [265]:
len(multiline_string)

3

In [266]:
print('a \nb')

a 
b


A string can be converted into a list of characters using `list` function.

In [267]:
list(multiline_string)

['a', '\n', 'b']

In [268]:
a_music_pun

'\nTwo windmills are standing in a field and one asks the other,\n"What kind of music do you like?"\n\nThe other says,\n"I\'m a big metal fan."\n'

In [269]:
print(a_music_pun)


Two windmills are standing in a field and one asks the other,
"What kind of music do you like?"

The other says,
"I'm a big metal fan."



In [270]:
list(a_music_pun)

['\n',
 'T',
 'w',
 'o',
 ' ',
 'w',
 'i',
 'n',
 'd',
 'm',
 'i',
 'l',
 'l',
 's',
 ' ',
 'a',
 'r',
 'e',
 ' ',
 's',
 't',
 'a',
 'n',
 'd',
 'i',
 'n',
 'g',
 ' ',
 'i',
 'n',
 ' ',
 'a',
 ' ',
 'f',
 'i',
 'e',
 'l',
 'd',
 ' ',
 'a',
 'n',
 'd',
 ' ',
 'o',
 'n',
 'e',
 ' ',
 'a',
 's',
 'k',
 's',
 ' ',
 't',
 'h',
 'e',
 ' ',
 'o',
 't',
 'h',
 'e',
 'r',
 ',',
 '\n',
 '"',
 'W',
 'h',
 'a',
 't',
 ' ',
 'k',
 'i',
 'n',
 'd',
 ' ',
 'o',
 'f',
 ' ',
 'm',
 'u',
 's',
 'i',
 'c',
 ' ',
 'd',
 'o',
 ' ',
 'y',
 'o',
 'u',
 ' ',
 'l',
 'i',
 'k',
 'e',
 '?',
 '"',
 '\n',
 '\n',
 'T',
 'h',
 'e',
 ' ',
 'o',
 't',
 'h',
 'e',
 'r',
 ' ',
 's',
 'a',
 'y',
 's',
 ',',
 '\n',
 '"',
 'I',
 "'",
 'm',
 ' ',
 'a',
 ' ',
 'b',
 'i',
 'g',
 ' ',
 'm',
 'e',
 't',
 'a',
 'l',
 ' ',
 'f',
 'a',
 'n',
 '.',
 '"',
 '\n']

Strings also support several list operations, which are discussed in the next section. We'll look at a couple of examples here.

#### Indexing and Selection

You can access individual characters within a string using the **`[]`** `indexing notation`. Note the character indices go from `0` to `n-1`, where `n` is the length of the string.

In [271]:
name = 'Jennifer'

In [272]:
name[3]

'n'

In [273]:
name[6]

'e'

In [274]:
name[7]

'r'

In [275]:
name[0]

'J'

##### Negative

In [276]:
name

'Jennifer'

In [277]:
name[-4]

'i'

In [278]:
name[-1]

'r'

In [279]:
name[-8]

'J'

In [280]:
df = 'Saturday'

In [281]:
df[-6]

't'

In [282]:
df[2]

't'

In [283]:
df[-4]

'r'

In [284]:
df[4]

'r'

#### Slicing/Selection

You can access a part of a string using by providing a `start:end:stepping` range instead of a single index in `[]`.

In [285]:
name

'Jennifer'

In [286]:
name[:4]

'Jenn'

In [287]:
name[3:]

'nifer'

In [288]:
name[::3]

'Jne'

In [289]:
name[::-2]

'rfne'

In [290]:
name

'Jennifer'

In [291]:
name[3:7]

'nife'

In [292]:
name[0:5]

'Jenni'

In [293]:
name

'Jennifer'

In [294]:
name[:6:2]

'Jni'

In [295]:
name[5:8]

'fer'

In [296]:
name[0:7:3]

'Jne'

In [297]:
name

'Jennifer'

In [298]:
name[-5:-1]

'nife'

In [299]:
name[1:-4]

'enn'

#### Membership

You can also check whether a string contains some text using the `in` operator.

In [300]:
today = 'Saturday'

In [301]:
today

'Saturday'

In [302]:
'Day' in today

False

In [303]:
'day' == 'Day'

False

In [304]:
'Sat' in today

True

In [305]:
'Wed' not in today

True

#### Concating Strings

Two or more strings can be joined or *concatenated* using the `+` operator.

**`Note`**: Be careful while concatenating strings, sometimes you may need to add a space character `" "` between words.

In [306]:
full_name = "Philip Eze"
full_name

'Philip Eze'

In [307]:
name = 'Philip'

surname = 'Eze'

In [312]:
name

'Philip'

In [313]:
surname

'Eze'

In [308]:
full_name_1 = name + surname
full_name_1

'PhilipEze'

In [309]:
full_name_1 = name + ' ' +surname
full_name_1

'Philip Eze'

#### Replicating Strings

In [316]:
# TypeError: Cannot concatenate a string and an integer directly using '+'.
# NameError: 'name' must also be defined in the current scope.
name + str(2)

'Philip2'

### Explanation: String Concatenation with Numbers

This code demonstrates how to safely combine a string with a number using **casting**:

1. **`name`**: This is a variable holding a string (e.g., `'Philip'`).
2. **`str(2)`**: The `str()` function converts the integer `2` into the string `'2'`.
3. **`+`**: Since both values are now strings, the `+` operator performs **concatenation**, joining them together into a single string.

**Note**: Without using `str()`, Python would raise a `TypeError` because it cannot directly add a number to text.

In [317]:
name = 'Philip'
name

'Philip'

In [318]:
name * 2

'PhilipPhilip'

In [319]:
name * 4

'PhilipPhilipPhilipPhilip'

In [320]:
name * 4

'PhilipPhilipPhilipPhilip'

In [324]:
# TypeError: Cannot multiply two strings together; multiplication requires an integer operand.
name *6

'PhilipPhilipPhilipPhilipPhilipPhilip'

### Explanation: TypeError in String Multiplication

This cell produces a `TypeError` because of how the multiplication operator `*` works with strings:

1. **Replication**: In Python, you can multiply a **string** by an **integer** (e.g., `'Philip' * 3`). This results in the string being repeated that many times (`'PhilipPhilipPhilip'`).
2. **The Error**: You **cannot** multiply a string by another string (`name * name`). Python does not have a defined mathematical or logical operation for 'squaring' text.

**To fix this**: Ensure that the second operand is an integer if you intend to repeat the text, or use addition (`+`) if you intended to join (concatenate) strings together.

In [325]:
name_5 = name * 5
name_5

'PhilipPhilipPhilipPhilipPhilip'

#### String Methods

Strings in Python have many built-in *methods* that are used to manipulate them. Let's try out some common string methods.

> **Methods**: Methods are functions associated with data types and are accessed using the `.` notation e.g. `variable_name.method()` or `"a string".method()`. Methods are a powerful technique for associating common operations with values of specific data types.

The `.lower()`, `.upper()` and `.capitalize()` methods are used to change the case of the characters.

In [326]:
today = "TUESDAY"

In [327]:
today

'TUESDAY'

In [328]:
today = today.lower()

In [329]:
today

'tuesday'

In [330]:
tomorrow = "Sunday"

In [331]:
tomorrow

'Sunday'

In [332]:
tomorrow.lower()

'sunday'

In [333]:
tomorrow

'Sunday'

In [334]:
today = 'saturday'
today

'saturday'

In [335]:
# NameError: 'today' must be defined before calling string methods like .upper().
today.upper()

'SATURDAY'

In [336]:
today

'saturday'

In [337]:
today = today.upper()
today

'SATURDAY'

In [342]:
today

'SATURDAY'

In [343]:
name = 'solomon'

In [338]:
name.capitalize()

'Philip'

In [339]:
name

'Philip'

In [340]:
name = name.capitalize()
name

'Philip'

In [341]:
"monday".capitalize() # changes first character to uppercase

'Monday'

The `.replace` method replaces a part of the string with another string. It takes the portion to be replaced and the replacement text as *inputs* or *arguments*.

In [344]:
today = "SATURDAY"

In [ ]:
today.replace('SATUR', 'WEDNES')

In [345]:
today

'SATURDAY'

In [346]:
another_day = today.replace("SATUR", "WEDNES")

In [347]:
another_day

'WEDNESDAY'

Note that `replace` returns a new string, and the original string is not modified.

In [348]:
today

'SATURDAY'

The `.split` method splits a string into a list of strings at every occurrence of provided character(s).

In [349]:
"Sun,Mon,Tue,Wed,Thu,Fri,Sat"

'Sun,Mon,Tue,Wed,Thu,Fri,Sat'

In [350]:
week = "Sun,Mon,Tue,Wed,Thu,Fri,Sat".split(',')
week

['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat']

In [351]:
"1|2|3|4|5|6|7".split("|")

['1', '2', '3', '4', '5', '6', '7']

The `.strip` method removes whitespace characters from the beginning and end of a string.

In [352]:
a_long_line = "       This is a long line with some space before, after,     and some space in the middle..    "

In [353]:
print(a_long_line)

       This is a long line with some space before, after,     and some space in the middle..    


In [354]:
a_long_line

'       This is a long line with some space before, after,     and some space in the middle..    '

In [355]:
a_long_line_stripped = a_long_line.strip()

In [356]:
a_long_line_stripped

'This is a long line with some space before, after,     and some space in the middle..'

The `.format` method combines values of other data types, e.g., integers, floats, booleans, lists, etc. with strings. You can use `format` to construct output messages for display.

In [363]:
# Fixed: Use curly braces as placeholders and call .format() on the string literal.
"I am Solomon and I am {} year old".format(27)

'I am Solomon and I am 27 year old'

In [359]:
"I am Solomon and I am {} year old".format(27)

'I am Solomon and I am 27 year old'

In [361]:
print('I  am  Solomon and I am {} year old'.format(27))

I  am  Solomon and I am 27 year old


In [360]:
# Input variables
cost_of_ice_bag = 1.25
profit_margin = .2
number_of_bags = 500

In [364]:
# Template for output message
output_template = """If a grocery store sells ice bags at $ {} per bag, with a profit margin of {} %,
then the total profit it makes by selling {} ice bags is $ {}."""

print(output_template)

If a grocery store sells ice bags at $ {} per bag, with a profit margin of {} %,
then the total profit it makes by selling {} ice bags is $ {}.


In [ ]:
print(output_template)

In [365]:
# Inserting values into the string

total_profit = cost_of_ice_bag * profit_margin * number_of_bags
total_profit

125.0

In [366]:
output_message = output_template.format(cost_of_ice_bag,
                                        profit_margin*100,
                                        number_of_bags, total_profit)

print(output_message)

If a grocery store sells ice bags at $ 1.25 per bag, with a profit margin of 20.0 %,
then the total profit it makes by selling 500 ice bags is $ 125.0.


Notice how the placeholders `{}` in the `output_template` string are replaced with the arguments provided to the `.format` method.

It is also possible to use the string concatenation operator `+` to combine strings with other values. However, those values must first be converted to strings using the `str` function.

In [368]:
# Fixed: Both float variables must be cast to strings using str() to allow concatenation.
"If a grocery store sells ice bags at $ " + str(cost_of_ice_bag) + ", with a profit margin of " + str(profit_margin * 100) + "%"

'If a grocery store sells ice bags at $ 1.25, with a profit margin of 20.0%'

In [369]:
"If a grocery store sells ice bags at $ " + str(cost_of_ice_bag) + ", with a profit margin of " + str(profit_margin*100)

'If a grocery store sells ice bags at $ 1.25, with a profit margin of 20.0'

#### Converting other data types to strings

You can us the  `str` function to convert a value of any data type into a string.

In [370]:
goh = 123
goh

123

In [371]:
goh = str(123) ## converting an integer to a string
goh

'123'

In [372]:
str(23.432) ## converting a floating value to a string

'23.432'

In [373]:
str(True)

'True'

In [374]:
the_3_musketeers = ["Athos", "Porthos", "Aramis"]
str(the_3_musketeers)

"['Athos', 'Porthos', 'Aramis']"

**Note**: All string methods return new values and DO NOT change the existing string. You can find a full list of string methods here: https://www.w3schools.com/python/python_ref_string.asp.

#### Comparism Operations on Strings

Strings also support the comparison operators `==` and `!=` for checking whether two strings are equal.

In [375]:
first_name = "Mary"

In [376]:
first_name == "Mariam"

False

In [377]:
first_name == "Mary"

True

In [378]:
# NameError: Comparison failed because 'first_name' was not defined.
first_name != "Jane"

True

We've looked at the primitive data types in Python. We're now ready to explore non-primitive data structures, also known as containers.

### List

`A list in Python is an ordered collection of values`.

Lists can hold values of different data types and support operations to `add`, `remove`, and `change values`. Lists have the type `list`.

To create a list, enclose a sequence of values within square brackets `[` and `]`, separated by commas.

In [379]:
sol =  ['Promise', 'Solomon']
solo = ['Promise', 'Solomon']

In [380]:
sol == solo

True

In [381]:
sol[1]

'Solomon'

In [382]:
solo[1]

'Solomon'

In [383]:
fruits = ['apple', 'banana', 'cherry']
fruits

['apple', 'banana', 'cherry']

In [387]:
fruits_i = ['banana','apple', 'cherry']
fruits_i

['banana', 'apple', 'cherry']

In [388]:
fruits == fruits_i

False

In [389]:
fruits_2 = ['apple', 'banana',  'cherry', 1, False, 45.99]

In [390]:
fruits_2

['apple', 'banana', 'cherry', 1, False, 45.99]

In [385]:
fruits

['apple', 'banana', 'cherry']

In [386]:
type(fruits)

list

Let's try creating a list containing values of different data types, including another list.

In [391]:
a_list = [23, 'hello', None, 3.14, fruits, 3 <= 5]

In [392]:
a_list

[23, 'hello', None, 3.14, ['apple', 'banana', 'cherry'], True]

In [393]:
empty_list = []

In [394]:
empty_list

[]

In [395]:
empty_list

[]

To determine the number of values in a list, use the `len` function. You can use `len`  to determine the number of values in several other data types.

In [396]:
# NameError: 'fruits' must be defined before it can be accessed.
fruits

['apple', 'banana', 'cherry']

In [ ]:
len(fruits)

In [397]:
print("Number of fruits:", len(fruits))

Number of fruits: 3


In [398]:
a_list

[23, 'hello', None, 3.14, ['apple', 'banana', 'cherry'], True]

In [399]:
len(a_list)

6

In [400]:
len(empty_list)

0

#### Indexing and Selection

You can access an element from the list using its *index*, e.g., `fruits[2]` returns the element at index 2 within the list `fruits`. The starting index of a list is 0.

In [401]:
fruits

['apple', 'banana', 'cherry']

In [402]:
fruits[0]

'apple'

In [403]:
fruits[1]

'banana'

In [404]:
fruits[2]

'cherry'

In [405]:
fruits[-1]

'cherry'

In [172]:
# NameError: The variable 'fruits' was not defined at this point in the execution when this cell was run.
fruits[-2]

'dates'

In [406]:
fruits[-3]

'apple'

If you try to access an index equal to or higher than the length of the list, Python returns an `IndexError`.

In [408]:
fruits

['apple', 'banana', 'cherry']

In [410]:
# IndexError: Index 3 is out of bounds for a list/tuple of length 3.
fruits[2]

'cherry'

In [411]:
# IndexError: The index `4` is out of range for the `fruits` list, which has 3 elements (indices 0, 1, 2).
fruits[0]

'apple'

You can use negative indices to access elements from the end of a list, e.g., `fruits[-1]` returns the last element, `fruits[-2]` returns the second last element, and so on.

In [412]:
fruits[-1]

'cherry'

In [ ]:
fruits[-2]

In [413]:
fruits[-3]

'apple'

In [ ]:
fruits[-4]

You can also access a range of values from the list. The result is itself a list. Let us look at some examples.

In [414]:
a_list = [23, 'hello', None, 3.14, fruits, 3 <= 5]

In [415]:
a_list

[23, 'hello', None, 3.14, ['apple', 'banana', 'cherry'], True]

In [416]:
len(a_list)

6

In [417]:
a_list[0]

23

In [418]:
a_list[1]

'hello'

In [419]:
a_list[2]

In [420]:
a_list[3]

3.14

In [421]:
a_list[4]

['apple', 'banana', 'cherry']

In [422]:
a_list[4][0]

'apple'

In [423]:
a_list[4][1]

'banana'

In [424]:
a_list[4][2]

'cherry'

In [ ]:
a_list[5]

In [425]:
a_list

[23, 'hello', None, 3.14, ['apple', 'banana', 'cherry'], True]

In [426]:
a_list[:4]

[23, 'hello', None, 3.14]

In [427]:
a_list[2:]

[None, 3.14, ['apple', 'banana', 'cherry'], True]

In [428]:
a_list[::2]

[23, None, ['apple', 'banana', 'cherry']]

In [429]:
a_list[0:3]

[23, 'hello', None]

In [438]:
# IndexError: You are trying to access index 10, but the list only has 6 elements (indices 0-5).
a_list[5]

True

In [431]:
a_list[2:6]

[None, 3.14, ['apple', 'banana', 'cherry'], True]

In [432]:
a_list[5:1]

[]

In [433]:
a_list

[23, 'hello', None, 3.14, ['apple', 'banana', 'cherry'], True]

In [434]:
a_list[2:] ### stopping at the last index (value)

[None, 3.14, ['apple', 'banana', 'cherry'], True]

In [435]:
a_list[:4]  ### Starting at the very begining

[23, 'hello', None, 3.14]

In [436]:
a_list

[23, 'hello', None, 3.14, ['apple', 'banana', 'cherry'], True]

In [437]:
 a_list[-3:]

[3.14, ['apple', 'banana', 'cherry'], True]

In [439]:
# Fix: Defining a_list locally to ensure it is recognized and explaining the slice.
a_list = [23, 'hello', None, 3.14, ['apple', 'banana', 'cherry'], True]
a_list[-2:-5]

[]

In [440]:
 a_list[5:2]

[]

Note that the range `2:5` includes the element at the start index `2` but does not include the element at the end index `5`. So, the result has 3 values (index `2`, `3`, and `4`).

Here are some experiments you should try out (use the empty cells below):

* Try setting one or both indices of the range are larger than the size of the list, e.g., `a_list[2:10]`
* Try setting the start index of the range to be larger than the end index, e.g., `a_list[12:10]`
* Try leaving out the start or end index of a range, e.g., `a_list[2:]` or `a_list[:5]`
* Try using negative indices for the range, e.g., `a_list[-2:-5]` or `a_list[-5:-2]` (can you explain the results?)

> The flexible and interactive nature of Jupyter notebooks makes them an excellent tool for learning and experimentation. If you are new to Python, you can resolve most questions as soon as they arise simply by typing the code into a cell and executing it. Let your curiosity run wild, discover what Python is capable of and what it isn't!

In [441]:
a_list[2:10]

[None, 3.14, ['apple', 'banana', 'cherry'], True]

In [442]:
a_list[12:10]

[]

In [443]:
a_list[2:]

[None, 3.14, ['apple', 'banana', 'cherry'], True]

#### Updating Values in a list

You can also change the value at a specific index within a list using the assignment operation.

In [444]:
fruits

['apple', 'banana', 'cherry']

In [445]:
fruits[1]

'banana'

In [446]:
fruits[1] = 'blueberry'

In [447]:
fruits

['apple', 'blueberry', 'cherry']

In [448]:
fruits[2] = 'mango'

In [449]:
fruits

['apple', 'blueberry', 'mango']

In [450]:
names = ['Seyi', 'Isaiah', 'Chris', 'Samuel', 'Sodiq']

In [451]:
names[3] = 1000
names

['Seyi', 'Isaiah', 'Chris', 1000, 'Sodiq']

In [452]:
names[2:4]  = ['Promise']

In [453]:
names

['Seyi', 'Isaiah', 'Promise', 'Sodiq']

#### List methods

A new value can be added to the end of a list using the `append` method.

In [454]:
# NameError: 'fruits' must be defined as a list or tuple before access.
fruits

['apple', 'blueberry', 'mango']

#### `.append()`

`Append one object at a time to the end of the list.`

It takes only one value at a time.

In [455]:
fruits.append('dates')

In [456]:
fruits

['apple', 'blueberry', 'mango', 'dates']

In [458]:
# Fix: .append() only takes one argument. To add multiple, use .extend().
fruits.extend(['guava', 'pawpaw'])
fruits

['apple', 'blueberry', 'mango', 'dates', 'guava', 'pawpaw']

In [459]:
fruits + ['guava', 'apple'] ## concating two list together !!!

['apple', 'blueberry', 'mango', 'dates', 'guava', 'pawpaw', 'guava', 'apple']

#### `insert`

A new value can also be inserted at a specific index using the `insert` method.

In [461]:
fruits

['apple', 'blueberry', 'mango', 'dates', 'guava', 'pawpaw']

In [462]:
fruits.insert(1, 'banana')
fruits

['apple', 'banana', 'blueberry', 'mango', 'dates', 'guava', 'pawpaw']

In [463]:
fruits.insert(4, 'pawpaw')
fruits

['apple', 'banana', 'blueberry', 'mango', 'pawpaw', 'dates', 'guava', 'pawpaw']

`Extend` method

In [467]:
fruits.extend(['groundnut', 'orange'])

In [468]:
fruits

['apple',
 'banana',
 'blueberry',
 'mango',
 'pawpaw',
 'dates',
 'guava',
 'pawpaw',
 'guava',
 'apple',
 'kiwi',
 'peach',
 'pineapple',
 'watermelon',
 'tingerine',
 'groundnut',
 'orange']

In [469]:
fruits.extend(['kiwi', 'peach' , 'pineapple', 'watermelon', 'tingerine'])
fruits

['apple',
 'banana',
 'blueberry',
 'mango',
 'pawpaw',
 'dates',
 'guava',
 'pawpaw',
 'guava',
 'apple',
 'kiwi',
 'peach',
 'pineapple',
 'watermelon',
 'tingerine',
 'groundnut',
 'orange',
 'kiwi',
 'peach',
 'pineapple',
 'watermelon',
 'tingerine']

#### `remove`

You can remove a value from a list using the `remove` method.

In [470]:
fruits

['apple',
 'banana',
 'blueberry',
 'mango',
 'pawpaw',
 'dates',
 'guava',
 'pawpaw',
 'guava',
 'apple',
 'kiwi',
 'peach',
 'pineapple',
 'watermelon',
 'tingerine',
 'groundnut',
 'orange',
 'kiwi',
 'peach',
 'pineapple',
 'watermelon',
 'tingerine']

In [471]:
fruits.remove('peach')
fruits

['apple',
 'banana',
 'blueberry',
 'mango',
 'pawpaw',
 'dates',
 'guava',
 'pawpaw',
 'guava',
 'apple',
 'kiwi',
 'pineapple',
 'watermelon',
 'tingerine',
 'groundnut',
 'orange',
 'kiwi',
 'peach',
 'pineapple',
 'watermelon',
 'tingerine']

In [473]:
# Fix: .remove() takes a single item. To remove multiple, call it multiple times.
fruits.remove('guava')
fruits.remove('apple')
fruits

['banana',
 'blueberry',
 'mango',
 'pawpaw',
 'dates',
 'pawpaw',
 'guava',
 'apple',
 'kiwi',
 'pineapple',
 'watermelon',
 'tingerine',
 'groundnut',
 'orange',
 'kiwi',
 'peach',
 'pineapple',
 'watermelon',
 'tingerine']

In [474]:
fruits

['banana',
 'blueberry',
 'mango',
 'pawpaw',
 'dates',
 'pawpaw',
 'guava',
 'apple',
 'kiwi',
 'pineapple',
 'watermelon',
 'tingerine',
 'groundnut',
 'orange',
 'kiwi',
 'peach',
 'pineapple',
 'watermelon',
 'tingerine']

In [476]:
# TypeError: .remove() expects one argument.
# AttributeError: Tuples do not support .remove().
fruits.remove('apple',)

What happens if a list has multiple instances of the value passed to `.remove`? Try it out.

In [155]:
# NameError: 'fruits' is not defined in the local/global scope.
fruits

('apple', 'cherry', 'dates', 'cherry')

In [477]:
fruits.remove('watermelon')

In [ ]:
fruits.remove('tingerine')

In [478]:
fruits

['banana',
 'blueberry',
 'mango',
 'pawpaw',
 'dates',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'orange',
 'kiwi',
 'peach',
 'pineapple',
 'watermelon',
 'tingerine']

#### `Pop`

To remove an element from a specific index, use the `pop` method. The method also returns the removed element.

In [483]:
fruits

['banana',
 'blueberry',
 'mango',
 'dates',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'orange',
 'kiwi',
 'peach',
 'pineapple',
 'watermelon',
 'tingerine']

In [484]:
fruits.pop(-4)

'peach'

In [479]:
fruits

['banana',
 'blueberry',
 'mango',
 'pawpaw',
 'dates',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'orange',
 'kiwi',
 'peach',
 'pineapple',
 'watermelon',
 'tingerine']

In [480]:
mango = fruits.pop(3)

In [481]:
mango

'pawpaw'

In [482]:
fruits

['banana',
 'blueberry',
 'mango',
 'dates',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'orange',
 'kiwi',
 'peach',
 'pineapple',
 'watermelon',
 'tingerine']

If no index is provided, the `pop` method removes the last element of the list.

In [485]:
# AttributeError: Tuples are immutable and do not have a .pop() method.
fruits.pop()

'tingerine'

In [494]:
fruits

['banana',
 'blueberry',
 'mango',
 'dates',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'orange',
 'kiwi',
 'pineapple']

In [495]:
fruits.insert(-1, 'pineapple')

In [496]:
fruits

['banana',
 'blueberry',
 'mango',
 'dates',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'orange',
 'kiwi',
 'pineapple',
 'pineapple']

In [497]:
popped = fruits.pop()

In [498]:
fruits

['banana',
 'blueberry',
 'mango',
 'dates',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'orange',
 'kiwi',
 'pineapple']

In [499]:
popped

'pineapple'

In [500]:
fruits_i = fruits.pop()

In [501]:
fruits_i

'pineapple'

#### Membership Test

You can test whether a list contains a value using the `in` operator.

In [502]:
fruits.pop(-2)

'orange'

In [503]:
# NameError: 'fruits' is not defined.
fruits

['banana',
 'blueberry',
 'mango',
 'dates',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'kiwi']

In [504]:
'pineapple'  not in fruits

False

In [505]:
'cherry' in fruits

False

In [506]:
'apple' in fruits

False

In [158]:
# NameError: 'fruits' is not defined.
'pawpaw' in fruits

False

#### Concate List

To combine two or more lists, use the `+` operator. This operation is also called *concatenation*.

In [507]:
fruits

['banana',
 'blueberry',
 'mango',
 'dates',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'kiwi']

In [508]:
['orange', 'tomato', 'ginger']

['orange', 'tomato', 'ginger']

In [509]:
['dates', 'guava']

['dates', 'guava']

In [510]:
more_fruits = fruits + ['orange', 'tomato', 'ginger']  + ['dates', 'guava']

In [511]:
more_fruits

['banana',
 'blueberry',
 'mango',
 'dates',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'kiwi',
 'orange',
 'tomato',
 'ginger',
 'dates',
 'guava']

In [512]:
more_fruits

['banana',
 'blueberry',
 'mango',
 'dates',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'kiwi',
 'orange',
 'tomato',
 'ginger',
 'dates',
 'guava']

#### Copy (Deep Copy) and View (Shallow Copy)

To create a copy of a list, use the `.copy()` method. Modifying the copied list does not affect the original.

In [513]:
more_fruits

['banana',
 'blueberry',
 'mango',
 'dates',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'kiwi',
 'orange',
 'tomato',
 'ginger',
 'dates',
 'guava']

In [514]:
# NameError: 'more_fruits' is not defined.
more_fruits_copied = more_fruits.copy()

In [515]:
more_fruits_copied

['banana',
 'blueberry',
 'mango',
 'dates',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'kiwi',
 'orange',
 'tomato',
 'ginger',
 'dates',
 'guava']

In [516]:
more_fruits

['banana',
 'blueberry',
 'mango',
 'dates',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'kiwi',
 'orange',
 'tomato',
 'ginger',
 'dates',
 'guava']

In [517]:
# NameError: 'more_fruits' is not defined.
more_fruits == more_fruits_copied

True

In [518]:
# Modify the copy
more_fruits_copied.remove('dates')

In [519]:
more_fruits_copied.pop()

'guava'

In [520]:
more_fruits_copied

['banana',
 'blueberry',
 'mango',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'kiwi',
 'orange',
 'tomato',
 'ginger',
 'dates']

In [521]:
# Original list remains unchanged
more_fruits

['banana',
 'blueberry',
 'mango',
 'dates',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'kiwi',
 'orange',
 'tomato',
 'ginger',
 'dates',
 'guava']

Note that you cannot create a copy of a list by simply creating a new variable using the assignment operator `=`. The new variable will point to the same list, and any modifications performed using either variable will affect the other.

In [522]:
more_fruits

['banana',
 'blueberry',
 'mango',
 'dates',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'kiwi',
 'orange',
 'tomato',
 'ginger',
 'dates',
 'guava']

In [523]:
more_fruits_view = more_fruits ### Copying both values and memory space.
more_fruits_view

['banana',
 'blueberry',
 'mango',
 'dates',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'kiwi',
 'orange',
 'tomato',
 'ginger',
 'dates',
 'guava']

In [524]:
more_fruits_view == more_fruits

True

In [525]:
more_fruits_view.remove('dates')
more_fruits_view.pop()

'guava'

In [526]:
# NameError: 'more_fruits_view' was not successfully created due to earlier errors.
more_fruits_view

['banana',
 'blueberry',
 'mango',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'kiwi',
 'orange',
 'tomato',
 'ginger',
 'dates']

In [527]:
more_fruits

['banana',
 'blueberry',
 'mango',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'kiwi',
 'orange',
 'tomato',
 'ginger',
 'dates']

In [538]:
# Fix: Check if the item exists before removing to avoid ValueError
if 'mango' in more_fruits:
    more_fruits.remove('mango')
else:
    print("'mango' not found in the list.")

more_fruits

'mango' not found in the list.


['banana',
 'blueberry',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'kiwi',
 'orange',
 'tomato',
 'ginger',
 'dates',
 'dates',
 'dates',
 'dates']

In [539]:
more_fruits.append('dates')

In [535]:
more_fruits

['banana',
 'blueberry',
 'pawpaw',
 'guava',
 'kiwi',
 'pineapple',
 'tingerine',
 'groundnut',
 'kiwi',
 'orange',
 'tomato',
 'ginger',
 'dates',
 'dates',
 'dates',
 'dates']

In [536]:
more_fruits_view[::-1]

['dates',
 'dates',
 'dates',
 'dates',
 'ginger',
 'tomato',
 'orange',
 'kiwi',
 'groundnut',
 'tingerine',
 'pineapple',
 'kiwi',
 'guava',
 'pawpaw',
 'blueberry',
 'banana']

Just like strings, there are several in-built methods to manipulate a list. However, unlike strings, most list methods modify the original list rather than returning a new one. Check out some common list operations here: https://www.w3schools.com/python/python_ref_list.asp .


Following are some exercises you can try out with list methods (use the blank code cells below):

* Reverse the order of elements in a list
* Add the elements of one list at the end of another list
* Sort a list of strings in alphabetical order
* Sort a list of numbers in decreasing order

In [541]:
### Exercise 1: Reverse the order of elements
more_fruits.reverse()
print("Reversed list:", more_fruits)

### Exercise 2: Add elements of one list to another
new_items = ['pomegranate', 'elderberry']
more_fruits.extend(new_items)
print("Extended list:", more_fruits)

### Exercise 3: Sort strings in alphabetical order
more_fruits.sort()
print("Alphabetical sort:", more_fruits)

### Exercise 4: Sort numbers in decreasing order
numbers = [45, 12, 89, 3, 102]
numbers.sort(reverse=True)
print("Decreasing numbers:", numbers)

Reversed list: ['dates', 'dates', 'dates', 'dates', 'dates', 'ginger', 'tomato', 'orange', 'kiwi', 'groundnut', 'tingerine', 'pineapple', 'kiwi', 'guava', 'pawpaw', 'blueberry', 'banana']
Extended list: ['dates', 'dates', 'dates', 'dates', 'dates', 'ginger', 'tomato', 'orange', 'kiwi', 'groundnut', 'tingerine', 'pineapple', 'kiwi', 'guava', 'pawpaw', 'blueberry', 'banana', 'pomegranate', 'elderberry']
Alphabetical sort: ['banana', 'blueberry', 'dates', 'dates', 'dates', 'dates', 'dates', 'elderberry', 'ginger', 'groundnut', 'guava', 'kiwi', 'kiwi', 'orange', 'pawpaw', 'pineapple', 'pomegranate', 'tingerine', 'tomato']
Decreasing numbers: [102, 89, 45, 12, 3]


### Tuple

A tuple is an ordered collection of values, similar to a list. However, it is not possible to add, remove, or modify values in a tuple. A tuple is created by enclosing values within parentheses `(` and `)`, separated by commas.

> Any data structure that cannot be modified after creation is called *immutable*. You can think of tuples as immutable lists.

Let's try some experiments with tuples.

In [102]:
fruits = ('apple', 'cherry', 'dates', 'cherry')
fruits

('apple', 'cherry', 'dates', 'cherry')

In [103]:
# check no. of elements
len(fruits)

4

In [104]:
# get an element (positive index)
fruits[0]

'apple'

In [105]:
# get an element (negative index)
fruits[-2]

'dates'

In [106]:
# check if it contains an element
'dates' in fruits

True

In [542]:
# TypeError: Tuples are immutable, meaning their elements cannot be changed after creation.
fruits[0] = 'avocado'

In [543]:
# AttributeError: Tuples are immutable and do not have an `append()` method to add new elements.
fruits.append('blueberry')

In [547]:
# To 'remove' an item from a tuple, convert it to a list first:
fruits_list = list(fruits)

# Check if 'apple' is in the list before trying to remove it
if 'apple' in fruits_list:
    fruits_list.remove('apple')
    print("'apple' removed.")
else:
    print("'apple' not found in the collection.")

fruits = tuple(fruits_list)
print(fruits)

'apple' not found in the collection.
('avocado', 'blueberry', 'mango', 'dates', 'pawpaw', 'guava', 'kiwi', 'pineapple', 'tingerine', 'groundnut', 'kiwi', 'blueberry')


In [549]:
# AttributeError: Tuples are immutable and do not have a `pop()` method.
# To fix this, convert to a list, pop the element, and convert back to a tuple.
fruits_list = list(fruits)
if fruits_list:
    popped_item = fruits_list.pop()
    fruits = tuple(fruits_list)
    print(f"Popped item: {popped_item}")
    print(f"Updated tuple: {fruits}")

Popped item: blueberry
Updated tuple: ('avocado', 'blueberry', 'mango', 'dates', 'pawpaw', 'guava', 'kiwi', 'pineapple', 'tingerine', 'groundnut', 'kiwi')


You can also skip the parantheses `(` and `)` while creating a tuple. Python automatically converts comma-separated values into a tuple.

In [551]:
the_3_musketeers = 'Athos', 'Porthos', 'Aramis'

In [552]:
the_3_musketeers

('Athos', 'Porthos', 'Aramis')

You can also create a tuple with just one element by typing a comma after it. Just wrapping it with parentheses `(` and `)` won't make it a tuple.

In [553]:
single_element_tuple = 4,

In [554]:
single_element_tuple

(4,)

In [555]:
another_single_element_tuple = (4,)

In [556]:
another_single_element_tuple

(4,)

In [557]:
not_a_tuple = (4)

In [558]:
not_a_tuple

4

Tuples are often used to create multiple variables with a single statement.

In [559]:
point = (3, 4, 5)

In [560]:
point_x, point_y, point_z = (3, 4, 5)

In [561]:
point_x

3

In [562]:
point_y

4

In [563]:
point_z

5

You can convert a list into a tuple using the `tuple` function, and vice versa using the `list` function

In [564]:
['one', 'two', 'three']

['one', 'two', 'three']

In [565]:
tuple(['one', 'two', 'three'])

('one', 'two', 'three')

In [566]:
list(('Athos', 'Porthos', 'Aramis'))

['Athos', 'Porthos', 'Aramis']

Tuples have just two built-in methods: `count` and `index`. Can you figure out what they do? While you look could look for documentation and examples online, there's an easier way to check a method's documentation, using the `help` function.

In [109]:
a_tuple = 23, "hello", False, None, 23, 37, "hello", 'hello'

In [110]:
a_tuple

(23, 'hello', False, None, 23, 37, 'hello', 'hello')

In [111]:
help(a_tuple.count)

Help on built-in function count:

count(value, /) method of builtins.tuple instance
    Return number of occurrences of value.



Within a Jupyter notebook, you can also start a code cell with `?` and type the name of a function or method. When you execute this cell, you will see the function/method's documentation in a pop-up window.

In [112]:
?a_tuple.index

Try using `count` and `index` with `a_tuple` in the code cells below.

In [113]:
a_tuple

(23, 'hello', False, None, 23, 37, 'hello', 'hello')

In [114]:
a_tuple.count(None)

1

In [115]:
a_tuple

(23, 'hello', False, None, 23, 37, 'hello', 'hello')

In [116]:
a_tuple.index('hello')

1

In [117]:
a_tuple.index(23)

0

In [118]:
a_tuple.index(23)

0

### Dictionary

A dictionary is an unordered collection of items. Each item stored in a dictionary has a key and value. You can use a key to retrieve the corresponding value from the dictionary.  Dictionaries have the type `dict`.

Dictionaries are often used to store many pieces of information e.g. details about a person, in a single variable. Dictionaries are created by enclosing key-value pairs within braces or curly brackets `{` and `}`.

In [119]:
person1 = {
    'name': 'John Doe',
    'sex': 'Male',
    'age': 32,
    'married': True,
}

In [120]:
person1

{'name': 'John Doe', 'sex': 'Male', 'age': 32, 'married': True}

In [567]:
print(person1)

{'name': 'John Doe', 'sex': 'Male', 'age': 32, 'married': True}


Dictionaries can also be created using the `dict` function.

In [122]:
person2 = dict(name='Jane Judy', sex='Female', age=28, married=False)

In [123]:
person2

{'name': 'Jane Judy', 'sex': 'Female', 'age': 28, 'married': False}

In [124]:
type(person1)

dict

Keys can be used to access values using square brackets `[` and `]`.

In [125]:
person1

{'name': 'John Doe', 'sex': 'Male', 'age': 32, 'married': True}

In [126]:
person1['name']

'John Doe'

In [127]:
person1['married']

True

In [128]:
person1['sex']

'Male'

In [129]:
person1['age']

32

In [130]:
person2

{'name': 'Jane Judy', 'sex': 'Female', 'age': 28, 'married': False}

In [131]:
person2['name']

'Jane Judy'

In [132]:
person2['age']

28

In [133]:
person2['sex']

'Female'

In [134]:
person2['married']

False

If a key isn't present in the dictionary, then a `KeyError` is *thrown*.

In [568]:
# KeyError: This happens because 'address' is not a key in the person1 dictionary.
# person1['address']

# To avoid the error, you can use the .get() method:
address = person1.get('address', 'Address not found')
print(address)

Address not found


You can also use the `get` method to access the value associated with a key.

In [ ]:
person2.get("name")

In [ ]:
person2

The `get` method also accepts a default value, returned if the key is not present in the dictionary.

In [ ]:
person2.get("address", "None")

You can check whether a key is present in a dictionary using the `in` operator.

In [ ]:
'name' in person1

In [ ]:
'address' in person1

You can change the value associated with a key using the assignment operator.

In [ ]:
person2['married']

In [ ]:
person2['married'] = True

In [ ]:
person2['married']

In [ ]:
person2

**The assignment operator can also be used to add new key-value pairs to the dictionary.**

In [ ]:
person1

In [ ]:
person1['address'] = '1, Penny Lane'

In [ ]:
person1

In [ ]:
person1['address']

In [ ]:
person1['email'] = 'johndoe42@gmail.com'

In [ ]:
person1['email']

In [ ]:
person1

To remove a key and the associated value from a dictionary, use the `pop` method.

In [ ]:
person1.pop('address')

In [ ]:
person1

Dictionaries also provide methods to view the list of keys, values, or key-value pairs inside it.

In [ ]:
person1.keys()

In [ ]:
person1.values()

In [ ]:
person1.items()

In [164]:
# TypeError: `dict.keys()` returns a dict_keys object, which is not subscriptable. Convert it to a list first if you need to access elements by index, e.g., `list(person1.keys())[1]`.
person1.keys()[1]

TypeError: 'dict_keys' object is not subscriptable

The results of `keys`, `values`, and `items` look like lists. However, they don't support the indexing operator `[]` for retrieving elements.

Can you figure out how to access an element at a specific index from these results? Try it below. *Hint: Use the `list` function*

Dictionaries provide many other methods. You can learn more about them here: https://www.w3schools.com/python/python_ref_dictionary.asp .

Here are some experiments you can try out with dictionaries (use the empty cells below):
* What happens if you use the same key multiple times while creating a dictionary?
* How can you create a copy of a dictionary (modifying the copy should not change the original)?
* Can the value associated with a key itself be a dictionary?
* How can you add the key-value pairs from one dictionary into another dictionary? Hint: See the `update` method.
* Can the dictionary's keys be something other than a string, e.g., a number, boolean, list, etc.?

In [ ]:
persons = {
    'name': ['John Doe','Solomon Promise', 'John Otor', 'Sarah Adams', 'Joyce Moses'],
    'sex': ['Male', 'Male', 'Male', 'Female', 'Female'],
    'age': [32,34, 24, 36, 25],
    'married': [True, False, False, True, True]
}

In [ ]:
persons

In [ ]:
persons['name']

In [ ]:
persons['name'][0]

In [ ]:
persons['name'].append('Akpan Felix')

In [ ]:
persons['age'].append(27)

In [ ]:
persons['name'][2]

In [ ]:
persons['name'][-1]

In [ ]:
persons['name'][4]

In [ ]:
persons['married'].append('True')

In [ ]:
persons['sex'].append('Male')

In [ ]:
persons

In [ ]:
persons['sex'].pop()

In [ ]:
print(persons)

In [ ]:
persons.keys()

In [ ]:
persons.values()

In [ ]:
persons.items()

## Further Reading

We've now completed our exploration of variables and common data types in Python. Following are some resources to learn more about data types in Python:

* Python official documentation: https://docs.python.org/3/tutorial/index.html
* Python Tutorial at W3Schools: https://www.w3schools.com/python/
* Practical Python Programming: https://dabeaz-course.github.io/practical-python/Notes/Contents.html

You are now ready to move on to the next tutorial: [Branching using conditional statements and loops in Python](https://jovian.ml/aakashns/python-branching-and-loops)

## Conditional Statements and Loops

Learning Objectives
By the end of this lesson, students should be able to:
1.	Understand decision-making in Python.
2.	Use comparison and logical operators.
3.	Write if, if-else, and if-elif-else statements.
4.	Use nested conditional statements.
5.	Understand and use for loops.
6.	Understand and use while loops.
7.	Use break and continue statements.
8.	Solve practical problems using loops and conditions.


# 1. Introduction to Conditional Statements
Conditional statements allow a program to make decisions.
Example:


In [ ]:
age = 18

if age >= 18:
    print("You are an adult")

# 2. Comparison Operators
these are some of the comparison Operators we will consider here:

==	Equal to
!=	Not equal to
 >	Greater than
 <	Less than
>=	Greater than or equal to
<=	Less than or equal to

In [ ]:
x = 10
y = 5

print(x > y)
print(x == y)

# Practice Exercise 1
Predict the output before running.

In [ ]:
a = 15
b = 20

print(a < b)
print(a >= b)
print(a != b)

# 3. The if Statement
Syntax:
if condition:
    statement



Example:

In [ ]:
score = 70

if score >= 50:
    print("Pass")


# Activity
Write a program that checks whether a number is positive.


In [ ]:
number = int(input("Enter a number: "))

if number > 0:
    print("Positive Number")

# 4. The if-else Statement
Syntax:
if condition:
    statement1
else:
    statement2

Example:

In [ ]:

age = 19

if age >= 18:
    print("You can vote")
else:
    print("You cannot vote")

# Exercise
Write a program that determines whether a number is even or odd.


In [ ]:
number = int(input("Enter a number: "))

if number % 2 == 0:
    print("Even")
else:
    print("Odd")

# 5. The if-elif-else Statement
Used when multiple conditions exist.

Example:

In [ ]:

score = 90

if score >= 70:
    print("Grade A")
elif score >= 60:
    print("Grade B")
elif score >= 50:
    print("Grade C")
else:
    print("Fail")

# Exercise
Create a grading system:
* 90–100 = A
* 80–89 = B
* 70–79 = C
* 60–69 = D
* Below 60 = F


In [ ]:
score = int(input("Enter score: "))

if score >= 90:
    print("A")
elif score >= 80:
    print("B")
elif score >= 70:
    print("C")
elif score >= 60:
    print("D")
else:
    print("F")


# 6. Logical Operators

Operator	Meaning
and	-> Both conditions must be True
or  -> At least one condition must be True
not	-> Reverses a condition


# Example:

In [ ]:
age = 25
citizen = True

if age >= 18 and citizen:
    print("Eligible to vote")

# Exercise
Create a login authentication system that check for user name and password

In [ ]:

username = "admin"
password = "1234"

if username == "admin" and password == "1234":
    print("Login Successful")
else:
    print("Login Failed")

# 7. Nested if Statements
An if statement inside another if statement.

Example:

In [ ]:

age = 20
has_id = True

if age >= 18:
    if has_id:
        print("Entry Granted")

# Practical Challenge
Write a program that checks the conditions for student to be promoted:
1.	Student must score at least 50.
2.	Attendance must be at least 75%.

In [ ]:
score = 65
attendance = 80

if score >= 50:
    if attendance >= 75:
        print("Promoted")
    else:
        print("Low Attendance")
else:
    print("Failed")


# 8. Introduction to Loops
Loops allow us to repeat tasks automatically.
Python has:
1.	for loop
2.	while loop



# The for Loop
Syntax:

for variable in sequence:
    statements



Example:

In [ ]:
for i in range(7):
    print(i)


# Understanding range()
* range(stop)
* range(start, stop)
* range(start, stop, step)


Examples:

In [ ]:

for i in range(1, 6):
    print(i)


# Counting Backward
Reversing the range


In [ ]:
for i in range(10, 0, -1):
    print(i)

# Exercise
Print all even numbers from 2 to 20.


In [ ]:
for i in range(2, 21, 2):
    print(i)


# 10. Looping Through Strings
Example:


In [ ]:
name = "PYTHON"

for letter in name:
    print(letter)

# Exercise
Print each character in your name.



In [ ]:
name = input("Enter your name: ")
for letter in name:
    print(letter)

# 11. The while Loop
Repeats as long as a condition remains True.
Syntax:
while condition:
    statements


Example:

In [ ]:
count = 1

while count <= 5:
    print(count)
    count += 1


# Exercise
Print numbers from 1 to 10.


In [ ]:
num = 1

while num <= 10:
    print(num)
    num += 1


# 12. Infinite Loops
Danger:


# 13. break Statement
Used to terminate a loop immediately.



Example:

In [ ]:
for i in range(10):
    if i == 5:
        break
    print(i)


# 14. Continue Statement
Skips the current iteration.


Example:

In [ ]:

for i in range(10):
    if i == 5:
        continue
    print(i)


# 15. Practical Examples
Example 1: Multiplication Table


In [ ]:
number = int(input("Enter a number: "))

for i in range(1, 13):
    print(number, "x", i, "=", number * i)


Example 2: Sum of Numbers

In [ ]:
total = 0

for i in range(1, 11):
    total += i

print("Sum =", total)


Example 3: Guessing Game

In [ ]:
secret = 7

guess = int(input("Guess the number: "))

while guess != secret:
    print("Try Again")
    guess = int(input("Guess again: "))

print("Correct!")


# Class Project
Student Result Checker
Requirements:
1.	Ask for student name.
2.	Ask for score.
3.	Display grade.
4.	Determine Pass/Fail.


# Assessment Questions
1.	What is the purpose of an if statement?
2.	Differentiate between if and if-else.
3.	What is a nested if statement?
4.	What is the difference between a for loop and a while loop?
5.	Explain break and continue.
6.	Write a program that prints numbers from 1 to 50.
7.	Write a program that prints all odd numbers from 1 to 100.
8.	Write a grading system using if-elif-else.


Thank you for joining us on this training we are interested in your journey in faith and also your journing in programming, we pray for greater heights for you in these journeys in Jesus name Amen.